File: backend/app/ingestion/orchestrator.py
"""Coordinates source isolation, candidate quotas, and Ingestion Run accounting."""

from __future__ import annotations

import uuid
from dataclasses import dataclass
from datetime import datetime, timedelta, timezone
from typing import Protocol

from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.db.models import Article, CategoryArticle, IngestionAttempt, IngestionRun
from app.ingestion.normalization import normalized_title_hash, url_hash
from app.ingestion.sources import MAX_CANDIDATES_PER_SOURCE, CandidateArticle


class SourceAdapter(Protocol):
    async def search(self, request: object) -> list[CandidateArticle]: ...


class IngestionStore(Protocol):
    async def mark_running(self, run_id: str) -> None: ...
    async def record_candidate(
        self,
        category_id: str,
        source_id: str,
        candidate: CandidateArticle,
    ) -> bool: ...
    async def record_attempt(self, category_id: str, source_id: str, status: str) -> None: ...
    async def finish(self, run_id: str, status: str, counts: "IngestionResult") -> None: ...


@dataclass(frozen=True)
class SourceWork:
    category_id: str
    source_id: str
    adapter: SourceAdapter


@dataclass(frozen=True)
class IngestionResult:
    candidate_count: int = 0
    inserted_count: int = 0
    duplicate_count: int = 0
    error_count: int = 0


class IngestionOrchestrator:
    def __init__(self, store: IngestionStore) -> None:
        self._store = store

    async def run_ingestion(self, run_id: str, work_items: list[SourceWork]) -> IngestionResult:
        await self._store.mark_running(run_id)
        counts = IngestionResult()
        for work in work_items:
            try:
                candidates = (await work.adapter.search(work))[:MAX_CANDIDATES_PER_SOURCE]
                inserted = 0
                duplicates = 0
                for candidate in candidates:
                    if await self._store.record_candidate(
                        work.category_id,
                        work.source_id,
                        candidate,
                    ):
                        inserted += 1
                    else:
                        duplicates += 1
                await self._store.record_attempt(work.category_id, work.source_id, "succeeded")
                counts = IngestionResult(
                    counts.candidate_count + len(candidates), counts.inserted_count + inserted,
                    counts.duplicate_count + duplicates, counts.error_count,
                )
            except Exception:
                await self._store.record_attempt(work.category_id, work.source_id, "failed")
                counts = IngestionResult(
                    counts.candidate_count, counts.inserted_count, counts.duplicate_count,
                    counts.error_count + 1,
                )
        await self._store.finish(run_id, "succeeded", counts)
        return counts


class SqlAlchemyIngestionStore:
    """Writes Articles, Category Articles, Attempts, and final Run counters in one session."""

    def __init__(self, session: AsyncSession) -> None:
        self._session = session
        self._run_id: str | None = None

    async def mark_running(self, run_id: str) -> None:
        run = await self._session.get(IngestionRun, uuid.UUID(run_id))
        if run is None:
            raise ValueError("Ingestion Run not found")
        if run.status == "queued":
            run.status = "running"
        self._run_id = run_id

    async def record_candidate(
        self,
        category_id: str,
        source_id: str,
        candidate: CandidateArticle,
    ) -> bool:
        canonical_hash = url_hash(candidate.canonical_url)
        title_hash = normalized_title_hash(candidate.title)
        article = await self._session.scalar(
            select(Article).where(Article.canonical_url_hash == canonical_hash)
        )
        if article is None:
            article = await self._session.scalar(
                select(Article).where(Article.normalized_title_hash == title_hash)
            )
        created = article is None
        if article is None:
            now = datetime.now(timezone.utc)
            article = Article(
                title=candidate.title,
                normalized_title_hash=title_hash,
                canonical_url=candidate.canonical_url,
                canonical_url_hash=canonical_hash,
                summary=candidate.summary,
                published_at=candidate.published_at,
                first_seen_at=now,
                expires_at=now + timedelta(days=30),
            )
            self._session.add(article)
            await self._session.flush()
        link = await self._session.get(
            CategoryArticle,
            {"category_id": uuid.UUID(category_id), "article_id": article.id},
        )
        if link is None:
            self._session.add(
                CategoryArticle(
                    category_id=uuid.UUID(category_id),
                    article_id=article.id,
                    source_setting_id=uuid.UUID(source_id),
                )
            )
        return created

    async def record_attempt(self, category_id: str, source_id: str, status: str) -> None:
        if self._run_id is None:
            raise RuntimeError("mark_running must be called before recording attempts")
        self._session.add(
            IngestionAttempt(
                run_id=uuid.UUID(self._run_id),
                category_id=uuid.UUID(category_id),
                source_setting_id=uuid.UUID(source_id),
                status=status,
            )
        )

    async def finish(self, run_id: str, status: str, counts: IngestionResult) -> None:
        run = await self._session.get(IngestionRun, uuid.UUID(run_id))
        if run is None:
            raise ValueError("Ingestion Run not found")
        run.status = status
        run.finished_at = datetime.now(timezone.utc)
        run.candidate_count = counts.candidate_count
        run.inserted_count = counts.inserted_count
        run.duplicate_count = counts.duplicate_count
        run.error_count = counts.error_count

File: backend/internal/ingestion/orchestrator.go
package ingestion

import (
	"context"
	"crypto/sha256"
	"database/sql"
	"encoding/hex"
	"errors"
	"strings"
	"time"

	"github.com/google/uuid"
)

const maxCandidatesPerSource = 10

// CandidateArticle is an adapter result before database deduplication.
type CandidateArticle struct {
	Title        string
	CanonicalURL string
	CitationURL  string
	Summary      *string
	PublishedAt  *time.Time
}

// SourceSearchResult separates source candidate accounting from eligible Articles.
// CandidateCount is capped source candidates considered before eligibility filtering.
type SourceSearchResult struct {
	Candidates     []CandidateArticle
	CandidateCount int
}

// SourceAdapter retrieves candidates for a single Source Setting.
type SourceAdapter interface {
	Search(context.Context, SourceWork) (SourceSearchResult, error)
}

// SourceWork binds one adapter to the Category and Source Setting it serves.
type SourceWork struct {
	CategoryID          uuid.UUID
	SourceID            uuid.UUID
	CategoryName        string
	ContentLanguage     string
	SearchKeywords      *string
	SpecialRequirements *string
	SourceLabel         string
	WebsiteInput        string
	Adapter             SourceAdapter
}

// Result is the final aggregate accounting for one Ingestion Run.
type Result struct {
	CandidateCount int
	InsertedCount  int
	DuplicateCount int
	ErrorCount     int
}

// Orchestrator isolates adapter failures while persisting each source atomically.
type Orchestrator struct {
	db  *sql.DB
	now func() time.Time
}

func NewOrchestrator(db *sql.DB) *Orchestrator {
	return &Orchestrator{db: db, now: time.Now}
}

func (orchestrator *Orchestrator) Run(ctx context.Context, runID uuid.UUID, work []SourceWork) (Result, error) {
	if err := orchestrator.markRunning(ctx, runID); err != nil {
		return Result{}, err
	}
	var result Result
	for _, item := range work {
		searchResult, err := item.Adapter.Search(ctx, item)
		if err != nil {
			if recordErr := orchestrator.recordFailure(ctx, runID, item, err); recordErr != nil {
				return Result{}, recordErr
			}
			result.ErrorCount++
			continue
		}
		searchResult = searchResult.capped()
		counts, err := orchestrator.recordSource(ctx, runID, item, searchResult)
		if err != nil {
			return Result{}, err
		}
		result.CandidateCount += counts.CandidateCount
		result.InsertedCount += counts.InsertedCount
		result.DuplicateCount += counts.DuplicateCount
	}
	if err := orchestrator.finish(ctx, runID); err != nil {
		return Result{}, err
	}
	return result, nil
}

func (orchestrator *Orchestrator) markRunning(ctx context.Context, runID uuid.UUID) error {
	result, err := orchestrator.db.ExecContext(ctx, `UPDATE ingestion_runs SET status='running' WHERE id=$1 AND status='queued'`, runID)
	if err != nil {
		return err
	}
	affected, err := result.RowsAffected()
	if err != nil {
		return err
	}
	if affected != 1 {
		return errors.New("ingestion run is not queued")
	}
	return nil
}

func (result SourceSearchResult) capped() SourceSearchResult {
	eligible := make([]CandidateArticle, 0, len(result.Candidates))
	for _, candidate := range result.Candidates {
		if isBlockedSourceURL(candidate.CanonicalURL) {
			continue
		}
		eligible = append(eligible, candidate)
	}
	result.Candidates = eligible
	if result.CandidateCount > maxCandidatesPerSource {
		result.CandidateCount = maxCandidatesPerSource
	}
	if len(result.Candidates) > result.CandidateCount {
		result.Candidates = result.Candidates[:result.CandidateCount]
	}
	return result
}

func (orchestrator *Orchestrator) recordSource(ctx context.Context, runID uuid.UUID, item SourceWork, searchResult SourceSearchResult) (Result, error) {
	tx, err := orchestrator.db.BeginTx(ctx, nil)
	if err != nil {
		return Result{}, err
	}
	defer tx.Rollback()
	result := Result{CandidateCount: searchResult.CandidateCount}
	for _, candidate := range searchResult.Candidates {
		created, err := orchestrator.recordCandidate(ctx, tx, item, candidate)
		if err != nil {
			return Result{}, err
		}
		if created {
			result.InsertedCount++
		} else {
			result.DuplicateCount++
		}
	}
	if _, err = tx.ExecContext(ctx, `INSERT INTO ingestion_attempts (id,run_id,category_id,source_setting_id,status,candidate_count,inserted_count,duplicate_count) VALUES ($1,$2,$3,$4,'succeeded',$5,$6,$7)`, uuid.New(), runID, item.CategoryID, item.SourceID, result.CandidateCount, result.InsertedCount, result.DuplicateCount); err != nil {
		return Result{}, err
	}
	if _, err = tx.ExecContext(ctx, `UPDATE ingestion_runs SET candidate_count=candidate_count+$2,inserted_count=inserted_count+$3,duplicate_count=duplicate_count+$4 WHERE id=$1`, runID, result.CandidateCount, result.InsertedCount, result.DuplicateCount); err != nil {
		return Result{}, err
	}
	if err = tx.Commit(); err != nil {
		return Result{}, err
	}
	return result, nil
}

func (orchestrator *Orchestrator) recordCandidate(ctx context.Context, tx *sql.Tx, item SourceWork, candidate CandidateArticle) (bool, error) {
	canonicalHash := fingerprint(candidate.CanonicalURL)
	titleHash := fingerprint(strings.ToLower(strings.TrimSpace(candidate.Title)))
	var articleID uuid.UUID
	var deletedAt *time.Time
	err := tx.QueryRowContext(ctx, `SELECT id,deleted_at FROM articles WHERE canonical_url_hash=$1`, canonicalHash).Scan(&articleID, &deletedAt)
	if errors.Is(err, sql.ErrNoRows) {
		err = tx.QueryRowContext(ctx, `SELECT id,deleted_at FROM articles WHERE normalized_title_hash=$1`, titleHash).Scan(&articleID, &deletedAt)
	}
	if err != nil && !errors.Is(err, sql.ErrNoRows) {
		return false, err
	}
	if deletedAt != nil {
		return false, nil
	}
	if errors.Is(err, sql.ErrNoRows) {
		articleID = uuid.New()
		now := orchestrator.now().UTC()
		expiresAt := now.AddDate(0, 0, 30)
		if _, err = tx.ExecContext(ctx, `INSERT INTO articles (id,title,normalized_title_hash,canonical_url,canonical_url_hash,citation_url,summary,published_at,first_seen_at,expires_at) VALUES ($1,$2,$3,$4,$5,$6,$7,$8,$9,$10)`, articleID, candidate.Title, titleHash, candidate.CanonicalURL, canonicalHash, candidate.CitationURL, candidate.Summary, candidate.PublishedAt, now, expiresAt); err != nil {
			return false, err
		}
		if _, err = tx.ExecContext(ctx, `INSERT INTO category_articles (category_id,article_id,source_setting_id) VALUES ($1,$2,$3) ON CONFLICT (category_id,article_id) DO NOTHING`, item.CategoryID, articleID, item.SourceID); err != nil {
			return false, err
		}
		return true, nil
	}
	if _, err = tx.ExecContext(ctx, `INSERT INTO category_articles (category_id,article_id,source_setting_id) VALUES ($1,$2,$3) ON CONFLICT (category_id,article_id) DO NOTHING`, item.CategoryID, articleID, item.SourceID); err != nil {
		return false, err
	}
	return false, nil
}

func (orchestrator *Orchestrator) recordFailure(ctx context.Context, runID uuid.UUID, item SourceWork, sourceErr error) error {
	tx, err := orchestrator.db.BeginTx(ctx, nil)
	if err != nil {
		return err
	}
	defer tx.Rollback()
	if _, err = tx.ExecContext(ctx, `INSERT INTO ingestion_attempts (id,run_id,category_id,source_setting_id,status,error_summary) VALUES ($1,$2,$3,$4,'failed',$5)`, uuid.New(), runID, item.CategoryID, item.SourceID, sourceErr.Error()); err != nil {
		return err
	}
	if _, err = tx.ExecContext(ctx, `UPDATE ingestion_runs SET error_count=error_count+1 WHERE id=$1`, runID); err != nil {
		return err
	}
	return tx.Commit()
}

func (orchestrator *Orchestrator) finish(ctx context.Context, runID uuid.UUID) error {
	_, err := orchestrator.db.ExecContext(ctx, `UPDATE ingestion_runs SET status='succeeded',finished_at=$2 WHERE id=$1 AND status='running'`, runID, orchestrator.now().UTC())
	return err
}

func fingerprint(value string) string {
	digest := sha256.Sum256([]byte(strings.TrimSpace(value)))
	return hex.EncodeToString(digest[:])
}
